### DISCUSSION TITLE
Can I get the Jacobian of a frame not defined in the URDF

### DISCUSSION THREAD

**Discussion by lwygzh** [stack-of-tasks/pinocchio#2138] [General]
Hi,

I have been using pinocchio to compute the Jacobians of the joints and frames of the robot that are defined in its URDF. 

I was wondering if it is possible to compute Jacobians that are not defined in the URDF, but has a fixed transformation with respect to one of the frames that are defined in the URDF.

Thanks!

**Comment by matheecs** [comment]
You can add a user defined frame by `addFrame()`.

**Reply by lwygzh** [reply]
@matheecs  I see, seems like a good solution!

Do you know if there is a way to remove this user-added frame when it is no longer in use?

**Reply by matheecs** [reply]
I think there is no method of API to remove the frame. Extra frames may not influence performance, so it is not necessary to remove them.

**Reply by jorisv** [reply]
Like say @matheecs there is no method to remove the frame.

If you don't want to compute the frame configuration anymore, a solution is to copy the model before calling `addFrame` and then, use the  old model when you don't need the frame anymore.

**Reply by lwygzh** [reply]
@jorisv @matheecs Thanks for the response!

I noticed one issue with the `addFrame` function. It does update the `robot.model` with the new frame, but I don't think it updates `robot.data`, so I can't even call `robot.data.oMf` with the new frame id. Do you know how to solve that?

**Reply by lwygzh** [reply]
I figured it out, just need to run

```python
robot.data = robot.model.createData()
```

and then update the pinocchio model with the joint angles and velocities.
